### ISSUE TITLE
Potential incompatibility with BSD-created pools

### ISSUE THREAD

**Issue by tannerdsilva** [zfsonlinux/zfs#6038]
### System information

Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | debian
Distribution Version    | jessie
Linux Kernel                 | linux-image-3.16.0-4-amd64
Architecture                 | x86_64
ZFS Version                  | 0.6.5.9-2~bpo8+1
SPL Version                  | 0.6.5.9-1~bpo8+1

### Describe the problem you're observing
I have a 5-drive raidz-1 pool that I created in FreeBSD **11** a few months ago. Recently, I decided to move to Debian for the more advanced virtualization technologies over bhyve. My pool operates EXACTLY as expected, except for when I try scrub the pool (containing over 1TB of data). The scrub will progress for about a minute or so, and then stop abruptly. This is the resulting message. I've rebooted the system and I've tried (to the best of my knowledge) to check compatibility between my pool and a pool I created in Linux. Everything seems to check out (feature flags match, etc). This issue is repeatable every time. There are no partitions on the disks.

```
  pool: tank
 state: ONLINE
  scan: scrub repaired 0 in 0h1m with 0 errors on Tue Apr 18 17:41:16 2017
config:

	NAME        STATE     READ WRITE CKSUM
	tank        ONLINE       0     0     0
	  raidz1-0  ONLINE       0     0     0
	    sda     ONLINE       0     0     0
	    sdb     ONLINE       0     0     0
	    sdc     ONLINE       0     0     0
	    sde     ONLINE       0     0     0
	    sdf     ONLINE       0     0     0

# zpool list tank
NAME   SIZE  ALLOC   FREE  EXPANDSZ   FRAG    CAP  DEDUP  HEALTH  ALTROOT
tank  13.6T  1.66T  12.0T         -     1%    12%  1.00x  ONLINE  -

```
The pool is otherwise completely functional. 

This issue is very similar to Issue #6011 (I was running FreeNAS as well), however, I created `tank` by hand. I didn't include partitions for swapping, as the FreeNAS utilities do by default. `tank` has all 5 disks, end-to-end, as shown by this snippet
```
# fdisk -l

Disk /dev/sdb: 2.7 TiB, 3000592982016 bytes, 5860533168 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disk /dev/sdc: 2.7 TiB, 3000592982016 bytes, 5860533168 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disk /dev/sde: 2.7 TiB, 3000592982016 bytes, 5860533168 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
```

### Describe how to reproduce the problem

In FreeBSD 11 `zpool create tank raidz1 /dev/ada0 ...`
Fill the pool with enough data to for a prolonged scrub
Import pool into a Debian Jessie machine running ZFS.
`zpool scrub tank`
Repeatedly issue `zpool status`, watch the scrub abruptly stop
### Include any warning/errors/backtraces from the system logs

I don't know how to pull more detailed information about the scrub. I'd be happy to provide more in depth logs if I knew where to find them.

**Comment by bunder2015** [comment]
possibly #5898 ?

**Comment by tannerdsilva** [comment]
Yes, probably the same root issue. I might add that my data (perhaps unlike #5898), is absolutely not mostly zeros.

**Comment by rincebrain** [comment]
If you found a way to reproduce this with a pool you could distribute disk images of, that'd be rather useful for debugging this, I imagine. :)

**Comment by loli10K** [comment]
Nice, this seems in fact to be the same problem. I don't think first issue linked here is going to be of any use since the pool is gone (manually destroyed) and the second one has a reproducer that doesn't work (for me): it would be nice if we could close those and keep the conversation in one place.

I'm still looking for a reproducer, so i guess i'm installing FreeBSD11, create a mirror (or raidz) bring it over to debian and see what happens.

**Comment by tannerdsilva** [comment]
I just tried recreating the issue with a single-disk zpool in VirtualBox (like an idiot, didn't even think about using a RAID). Contained one child dataset, no snapshots (probably should have done a few of these too). No issues with the scrub on this one. Maybe something in the RAID metadata is causing issues? I will continue to search for the reproducible path and report back...

**Comment by loli10K** [comment]
> Maybe something in the RAID metadata is causing issues?

The issue reported in https://github.com/zfsonlinux/zfs/issues/6011 is affecting a mirror, so i tried to reproduce it on a (very) small pool of mirrored virtual disk filled with random data:

```
  pool: freebsd11
 state: ONLINE
status: Some supported features are not enabled on the pool. The pool can
	still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
	the pool may no longer be accessible by software that does not support
	the features. See zpool-features(5) for details.
  scan: scrub in progress since Wed Apr 19 01:33:18 2017
	80MB scanned out of 447MB at 6.1MB/s, 0h0m to go
	0B repaired, 17.81% done
config:

	NAME        STATE     READ WRITE CKSUM
	freebsd11   ONLINE       0     0     0
	  mirror-0  ONLINE       0     0     0
	    vda     ONLINE       0     0     0
	    vdb     ONLINE       0     0     0

errors: No known data errors
```

Seems to be working as expected? This in on almost HEAD, i need to try this on 0.6.5.9.

@tannerdsilva it would be nice if you could:

1.  `#echo 1 > /sys/module/zfs/parameters/zfs_dbgmsg_enable`
2. scrub the pool (wait for it to stop)
3. post a gist of `cat /proc/spl/kstat/zfs/dbgmsg`

Thanks!

**Comment by tannerdsilva** [comment]
I'm sorry, I spent a good 20 minutes staring at this log (and comparing it against a scrub I issued to my SSD pool for reference), and couldn't find any weird discrepancies anywhere. To my eyes, a successful scrub and a failed scrub look roughly the same. It appears to be crawling the entire topology of my pool too, no dataset went "unscanned", which is also very strange.

I've included a sizable amount of the failed scrub log in hopes you can find the issue I'm not seeing.

```
1492583706   command: zpool scrub tank
1492583706   doing scan sync txg 1088886; bm=0/142/0/1
1492583706   resuming at 0/8e/0/1
1492583706   txg=3272 quiesce_txg=3273 sync_txg=3248
1492583706   quiesce done, handing off txg 3272
1492583706   txg=3272 quiesce_txg=3273 sync_txg=3248
1492583706   waiting; tx_synced=3272 waiting=3248 dp=ffff880458b11000
1492583709   txg=17342 quiesce_txg=17343 sync_txg=17317
1492583709   quiesce done, handing off txg 17342
1492583709   txg=17342 quiesce_txg=17343 sync_txg=17317
1492583709   waiting; tx_synced=17342 waiting=17317 dp=ffff880084b04800
1492583711   pausing at bookmark 0/4bc4/0/1
1492583711   pausing at DDT bookmark 3/0/0/0
1492583711   visited 2042 blocks in 5011ms
1492583711   ds=          (null) obj=b num=1
1492583711   ds=          (null) obj=a7 num=1
1492583711   ds=          (null) obj=1f num=1
1492583711   ds=          (null) obj=1f num=1
1492583711   txg=1088887 quiesce_txg=1088888 sync_txg=1088884
1492583711   quiesce done, handing off txg 1088887
1492583711   txg=1088887 quiesce_txg=1088888 sync_txg=1088884
1492583711   doing scan sync txg 1088887; bm=0/19396/0/1
1492583711   resuming at 0/4bc4/0/1
1492583711   scanned dataset 18 (tank/$ORIGIN@$ORIGIN) with min=0 max=1088884; pausing=0
1492583711   scanned dataset 21 (tank) with min=1 max=1088884; pausing=0
1492583711   txg=3273 quiesce_txg=3274 sync_txg=3248
1492583711   quiesce done, handing off txg 3273
1492583711   txg=3273 quiesce_txg=3274 sync_txg=3248
1492583711   waiting; tx_synced=3273 waiting=3248 dp=ffff880458b11000
1492583711   scanned dataset 1303 (tank/backups/stank/.vm_cache/boot2docker/initrd@backup_2017-04-14_11:14:55) with min=1 max=1028917; pausing=0
1492583711   scanned dataset 15 (tank/$ORIGIN) with min=1 max=1088884; pausing=0
1492583711   scanned dataset 1343 (tank/backups/stank/.vm_cache/boot2docker/initrd_old1@backup_2017-04-14_11:14:55) with min=1 max=1028925; pausing=0
1492583714   txg=17343 quiesce_txg=17344 sync_txg=17317
1492583714   quiesce done, handing off txg 17343
1492583714   txg=17343 quiesce_txg=17344 sync_txg=17317
1492583714   waiting; tx_synced=17343 waiting=17317 dp=ffff880084b04800
1492583716   pausing at bookmark 3d/442/0/21
1492583716   pausing at DDT bookmark 3/0/0/0
1492583716   scanned dataset 61 (tank/users/lightroom@pre-fix) with min=1 max=1079794; pausing=1
1492583716   visited 12111 blocks in 5015ms
1492583716   ds=          (null) obj=b num=1
1492583716   ds=          (null) obj=a7 num=1
1492583716   ds=          (null) obj=1f num=1
1492583716   ds=          (null) obj=1f num=1
1492583716   txg=1088888 quiesce_txg=1088889 sync_txg=1088884
1492583716   quiesce done, handing off txg 1088888
1492583716   txg=1088888 quiesce_txg=1088889 sync_txg=1088884
1492583716   doing scan sync txg 1088888; bm=61/1090/0/33
1492583716   scanned dataset 61 (tank/users/lightroom@pre-fix) with min=1 max=1079794; pausing=0
1492583716   scanned dataset 1293 (tank/backups/stank/.vm_cache/boot2docker/initrd_old0@backup_2017-04-14_11:14:55) with min=1 max=1028915; pausing=0
1492583716   scanned dataset 1203 (tank/backups/stank/.vm_cache/ubuntu-server-16.04-vnc/os@backup_2017-04-14_11:14:55) with min=1 max=1028897; pausing=0
1492583716   scanned dataset 1712 (tank/backups) with min=1 max=1088884; pausing=0
```
...these messages repeat in this fashion until...

```
1492583808   scanned dataset 1396 (tank/users/tannerdsilva@short-20170414.1600) with min=1026239 max=1029478; pausing=0
1492583808   scanned dataset 511 (tank/users/tannerdsilva@short-20170414.2000) with min=1029478 max=1032261; pausing=0
1492583808   scanned dataset 611 (tank/users/tannerdsilva@short-20170415.0000) with min=1032261 max=1035188; pausing=0
1492583808   scanned dataset 619 (tank/users/tannerdsilva@short-20170415.0400) with min=1035188 max=1038062; pausing=0
1492583809   scanned dataset 626 (tank/users/tannerdsilva@short-20170415.0800) with min=1038062 max=1040946; pausing=0
1492583809   scanned dataset 641 (tank/users/tannerdsilva@short-20170415.1200) with min=1040946 max=1043821; pausing=0
1492583809   scanned dataset 667 (tank/users/tannerdsilva@short-20170415.1600) with min=1043821 max=1046697; pausing=0
1492583809   txg=17362 quiesce_txg=17363 sync_txg=17317
1492583809   quiesce done, handing off txg 17362
1492583809   txg=17362 quiesce_txg=17363 sync_txg=17317
1492583809   waiting; tx_synced=17362 waiting=17317 dp=ffff880084b04800
1492583809   scanned dataset 675 (tank/users/tannerdsilva@short-20170415.2000) with min=1046697 max=1049576; pausing=0
1492583810   scanned dataset 680 (tank/users/tannerdsilva@short-20170416.0000) with min=1049576 max=1052451; pausing=0
1492583810   scanned dataset 685 (tank/users/tannerdsilva@short-20170416.0400) with min=1052451 max=1055328; pausing=0
1492583810   scanned dataset 695 (tank/users/tannerdsilva@short-20170416.0800) with min=1055328 max=1058188; pausing=0
1492583810   scanned dataset 701 (tank/users/tannerdsilva@short-20170416.1200) with min=1058188 max=1061054; pausing=0
1492583811   scanned dataset 55 (tank/users/tannerdsilva) with min=1061054 max=1088884; pausing=0
1492583811   txg=3293 quiesce_txg=3294 sync_txg=3248
1492583811   quiesce done, handing off txg 3293
1492583811   txg=3293 quiesce_txg=3294 sync_txg=3248
1492583811   waiting; tx_synced=3293 waiting=3248 dp=ffff880458b11000
1492583811   visited 31784 blocks in 3604ms
1492583811   txg 1088905 traversal complete, waiting till txg 1088906
1492583811   ds=          (null) obj=b num=1
1492583811   ds=          (null) obj=a7 num=1
1492583811   ds=          (null) obj=1f num=1
1492583811   ds=          (null) obj=1f num=1
1492583812   txg=1088906 quiesce_txg=1088907 sync_txg=1088884
1492583812   txg=1088906 quiesce_txg=1088907 sync_txg=1088884
1492583812   txg 1088906 scan complete
1492583812   dn=ffff880081220720 txg=1088906
1492583812   txg 1088906 scan done complete=1
1492583812   spa=tank async request task=8
1492583812   ds=          (null) obj=5c num=1
1492583812   ds=          (null) obj=b num=1
1492583812   ds=          (null) obj=a7 num=1
1492583812   ds=          (null) obj=14e num=1
1492583812   ds=          (null) obj=b6 num=1
1492583812   ds=          (null) obj=b7 num=1
1492583812   ds=          (null) obj=b8 num=1
1492583812   ds=          (null) obj=b9 num=1
1492583812   ds=          (null) obj=1f num=1
1492583812   ds=          (null) obj=1f num=1
1492583812   waiting; tx_synced=1088906 waiting=1088884 dp=ffff8804590a7800
1492583814   txg=17363 quiesce_txg=17364 sync_txg=17317
1492583814   quiesce done, handing off txg 17363
1492583814   txg=17363 quiesce_txg=17364 sync_txg=17317
1492583814   os=ffff88045fb6a000 obj=122, increase to 2
```

**Comment by loli10K** [comment]
@tannerdsilva how many different pools are imported on this system? 3? The log is quite confusing to me. Also i think you forgot to post the first part of the log, i expected to see the following line:

`txg 1088884 scan setup func=1 mintxg=0 maxtxg=1088884`

Finally, the reason i was asking for a gist is because with a truncated log we can't do things like `grep 'visited X blocks in Yms' $gist.file`.

I'm going to try and reproduce this on a virtual machine with a single pool and ZFS 0.6.5.9.

**Comment by tannerdsilva** [comment]
@loli10K My apologies, didn't realize a gist was a method of sharing snippets of code. This is my first real contribution on GitHub, still learning these things. :0

https://gist.github.com/tannerdsilva/7b26cbdf085343091b0eb11cdc20a095#file-gistfile1-txt

**Comment by dweeezil** [comment]
Someone with a working BSD pool should run "zdb -dddd <pool> 1" and make sure that the "scan" zap entry has the same number of values as it does under ZoL.  We (ZoL) had some issues in the past where it was modified inadvertently which led to the whole "errata fiasco".

**Comment by loli10K** [comment]
@dweeezil seems to be the same number?

FreeBSD11 (created a couple of hours ago to reproduce this issue, has never seen any other system):

```
[root@ ~]# zdb -dddd freebsd11 1 | grep 'scan = '
		scan = 1 3 0 0 148 1 148 1492637222 1492637702 12895930368 853191680 0 0 0 1 3 0 0 0 113 8 0 4320 0 
```

ZoL 0.7.0-rc3:

```
root@debian-8-zfs:~# zdb -dddd zol 1 | grep 'scan = '
		scan = 1 2 0 0 25 1 25 1492629519 1492629519 438272 146432 0 0 0 1 3 0 0 0 0 0 0 0 0 
```

**Comment by dweeezil** [comment]
@loli10K Does this problem happen if you migrate a FreeBSD pool on which a scrub has never been run?  It looks like your FreeBSD pool had an interrupted scrub.  It might also be useful to see the scrub/scan-related stuff from "zdb -hh" (internal history).

**Comment by loli10K** [comment]
>  It looks like your FreeBSD pool had an interrupted scrub

That would be my fault, `zdb` was taking too long so i stopped the scrub it manually.

> Does this problem happen if you migrate a FreeBSD pool on which a scrub has never been run?

I don't know, i haven't be able to reproduce this yet.

**Comment by tannerdsilva** [comment]
Not sure how helpful these datapoints are, but...
- 75,000 files currently in the refer across all datasets
- 31 datasets, some are replications from other datasets
- My old box was running an automatic snapshot utility (FreeNAS Corrall automated snapshots), currently has 111 snapshots
- One of my drives failed a few months ago, one of the devices has been resilvered

In theory, I wouldn't imagine this would make a difference, but clearly I've done something with my pool that we're not considering in the tests (I too am having a hard time reproducing the problem)

`zpool history tank` linked:
https://gist.github.com/tannerdsilva/c3e8b3a66b8223c0fc85dd0e182d99e9#file-tank_history-log

**Comment by tannerdsilva** [comment]
It doesn't look like I can even send `tank` to other pools. I created a test pool `tester` with some spare SAS drives on my physical box, issued `zfs send -R tank@clone | zfs recv -F tester`, the process breaks (but never exits) at the same point every time. In another ssh session, I can see `tester` only gets up to the third dataset.

I've tried this twice and got the same behavior both times

**Comment by tannerdsilva** [comment]
Yep, looks like I've narrowed down the problem to a single dataset in `tank`. The dataset **isn't** present in the `/proc/spl/kstat/zfs/dbgmsg` logs I posted earlier, so I'm pretty confident this is it (although, interestingly, the child datasets of this problematic dataset **are** in the log as "successfully scrubbed")

The problematic dataset is a essentially a backup and/or clone of another zfs dataset on my system, `stank`, and was backed up infrequently via block diffs using `zfs send | zfs recv` when I was on BSD

**Comment by loli10K** [comment]
If that's the case i'm still going to consider this a bug since `zpool status -v` should report some kind of error instead of "_scrub repaired 0 in 0h1m with 0 errors_".

> the process breaks (but never exits) at the same point every time
> The dataset isn't present in the /proc/spl/kstat/zfs/dbgmsg logs I posted earlier

@tannerdsilva to get to the bottom of this we need way more information:

1. Name of $problematic_dataset
2. Instead of piping `zfs send -R tank@clone | zfs recv -F tester` do it separately and on the single $problematic_dataset, so we at least know if it's `zfs send` or `zfs recv` that "breaks": `zfs send ... > /path/to/snapshot.dat` and then `zfs recv ... < /path/to/snapshot.dat`
3. When send/recv "breaks" check if `dmesg` shows anything ZFS related: if it does, post full output (use https://gist.github.com/)
4. Does `zpool status tank -v` report anything at all about this dataset? Post full output
5. Post a gist of `zpool events -v` and `zpool history -i` full outputs, **after** "send/recv" breaks

**Comment by tannerdsilva** [comment]
I'm not experiencing an issue when I follow the steps you describe with the single suspect dataset (non recursively). When I send the dataset with a recursive flag, thats when I get the expected behavior.

1. `tank/backups/stank` (and I guess...all child datasets as well)
2. `recv` breaks, `send` functions properly
3. Nothing.
4. Output after `recv` hang:
```
root@zaden:/home/tannerdsilva# zpool status tank -v
  pool: tank
 state: ONLINE
  scan: scrub repaired 0 in 0h1m with 0 errors on Wed Apr 19 02:36:52 2017
config:

	NAME        STATE     READ WRITE CKSUM
	tank        ONLINE       0     0     0
	  raidz1-0  ONLINE       0     0     0
	    sda     ONLINE       0     0     0
	    sdb     ONLINE       0     0     0
	    sdc     ONLINE       0     0     0
	    sde     ONLINE       0     0     0
	    sdf     ONLINE       0     0     0

errors: No known data errors
root@zaden:/home/tannerdsilva# 
```
5.  History:
https://gist.github.com/tannerdsilva/b9ccae3ba15a4422a663c6b2425bbe7e#file-zfs-history

Events:
https://gist.github.com/tannerdsilva/b9ccae3ba15a4422a663c6b2425bbe7e#file-zfs-events

**Comment by loli10K** [comment]
> the process breaks (but never exits)

> 3. Nothing.

@tannerdsilva so how does exactly `zfs recv` break? It just hangs? For how long?

Just out of curiosity, what's the uptime of this machine?

**Comment by tannerdsilva** [comment]
@loli10K It just hangs indefinitely, pretty much as if its still receiving data, but...

1. The HDD access light on my box goes from solid on to off and
2. I can see the receiving filesystem keeps a constant size from another SSH session. It gets about 6GB before dying.

Once I am confident process has hung, I can bash on ^C all day, but no response.

Woke up this morning, just checked my uptime, saw a suspicious load average...`1.00, 1,00, 1.00`. 
Further investigation with reveals the culprit:
```
top - 10:36:18 up  7:27,  1 user,  load average: 1.00, 1.00, 1.00
Tasks: 1258 total,   2 running, 1256 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.0 us, 12.6 sy,  0.0 ni, 87.4 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem:  16413724 total,  2687424 used, 13726300 free,     1044 buffers
KiB Swap:  2097148 total,        0 used,  2097148 free.    83872 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND    
 5341 root      20   0   35360   3264   2984 R  99.9  0.0 443:40.85 zfs        
 5551 tannerd+  20   0   26668   4004   2520 R   0.3  0.0   0:01.17 top        
    1 root      20   0   29016   5412   3132 S   0.0  0.0   0:02.85 systemd    
    2 root      20   0       0      0      0 S   0.0  0.0   0:00.02 kthreadd   
    3 root      20   0       0      0      0 S   0.0  0.0   0:00.08 ksoftirqd/0
```
I'm running a current generation Xeon on my box, it takes quite a bit of effort to consume a full core like this. Would it be useful for you if I truncated my `zfs recv` file to about 10GB and posted it here? It seems to consistently break at roughly 6GB into the process.

**Comment by loli10K** [comment]
> Would it be useful for you if I truncated my zfs recv file to about 10GB and posted it here?

No

> Once I am confident process has hung, I can bash on ^C all day, but no response.

I would have expected something, after a couple of minutes, from `demsg` output in this case, but i guess i was wrong.

Can you post the output of `cat /proc/${pid_of_blocked_zfs}/stack`?

**Comment by tannerdsilva** [comment]
Yep just checked `dmesg` for the third time. Definitely nothing zfs related.

Stack:
```
tannerdsilva@zaden:~$ sudo cat /proc/5341/stack
[sudo] password for tannerdsilva: 
[<ffffffff81192376>] kmem_cache_alloc_node_trace+0x476/0x4a0
[<ffffffffa02ae86d>] dbuf_find+0xed/0x130 [zfs]
[<ffffffffa02b16c9>] __dbuf_hold_impl+0xb9/0x4e0 [zfs]
[<ffffffffa00dfabf>] spl_kmem_zalloc+0xaf/0x160 [spl]
[<ffffffffa02b1b68>] dbuf_hold_impl+0x78/0x90 [zfs]
[<ffffffffa02c94af>] dnode_next_offset_level+0xdf/0x3f0 [zfs]
[<ffffffffa02cb755>] dnode_next_offset+0xc5/0x230 [zfs]
[<ffffffffa02bb37e>] dmu_object_next+0x3e/0x50 [zfs]
[<ffffffffa02c2b67>] restore_freeobjects.isra.8+0x57/0xa0 [zfs]
[<ffffffffa02c3bab>] dmu_recv_stream+0x71b/0xb00 [zfs]
[<ffffffffa0205521>] nvlist_common.part.102+0xe1/0x1e0 [znvpair]
[<ffffffffa032e325>] zfs_ioc_recv+0x1f5/0xb00 [zfs]
[<ffffffffa02decda>] dsl_prop_get_dd+0x18a/0x240 [zfs]
[<ffffffffa00e717f>] tsd_hash_dtor+0x6f/0x80 [spl]
[<ffffffffa00e7726>] tsd_set+0x136/0x400 [spl]
[<ffffffffa02def0c>] dsl_prop_get_ds+0x17c/0x230 [zfs]
[<ffffffffa00df876>] strdup+0x36/0x60 [spl]
[<ffffffffa02ee37d>] rrw_exit+0x3d/0x150 [zfs]
[<ffffffffa032d149>] zfsdev_ioctl+0x4a9/0x4e0 [zfs]
[<ffffffff811bceef>] do_vfs_ioctl+0x2cf/0x4b0
[<ffffffff811aa835>] vfs_read+0x145/0x170
[<ffffffff811bd151>] SyS_ioctl+0x81/0xa0
[<ffffffff8151adcd>] system_call_fast_compare_end+0x10/0x15
[<ffffffffffffffff>] 0xffffffffffffffff
tannerdsilva@zaden:~$ 
```

**Comment by loli10K** [comment]
@tannerdsilva we have seen an issue similar to this, but it should not happen when both `zfs send` and `zfs recv` use the same stable zfs-0.6.5.9.

One last thing you could try is to receive the stream into a different "test" pool on FreeBSD11 and ZoL 0.7.0 to see what happens.

**Comment by mrenouf** [comment]
Just to add, I'm experiencing similar symptoms from the same situation (migrated a pool created on FreeBSD 11, FreeNAS).

In my case everything seemed to work fine but I had a drive fail. When I try to resolver with a replacement, it hits a particular spot, slows for and then shockingly claims the resilver is complete (after only 3min).

Details here:
https://askubuntu.com/q/907003/1038

**Comment by loli10K** [comment]
@mrenouf can you post here a gist (https://gist.github.com) of your `zpool history -i $poolname`?

I'm seeing a trend here of disks getting replaced that i did not account for in my current setup to reproduce this, i'd like to know when/how the device was replaced so i can add this to my testing. Thanks!

**Comment by tannerdsilva** [comment]
@loli10K I tried to capture the the messages into a file, but couldn't get it for whatever reason.
Here's a picture of the resulting output when I take the `zfs send` file (generated in my Debian environment) and put it through `zfs recv` on BSD11.

![img_0409](https://cloud.githubusercontent.com/assets/1372214/25308093/6fd36a08-277b-11e7-8b0a-bc0c86bc9f72.JPG)

**Comment by loli10K** [comment]
@tannerdsilva that seems to be an unrelated issue with device name handling in FreeBSD. I suggest you try ZoL 0.7.0.

**Comment by tannerdsilva** [comment]
@loli10K same behavior on ZoL 0.7.0-rc3 as we saw in 0.6.5.9 stable. Hung process with 100% utilization. No exit.

I made sure I was running 0.7.0 using `modinfo zfs | grep -iw version` and `modinfo spl | grep -iw version`

**Comment by tannerdsilva** [comment]
I ran a scrub when I was booted in BSD earlier today, just for kicks. Absolutely no reported issues.

**Comment by loli10K** [comment]
> I ran a scrub when I was booted in BSD earlier today, just for kicks. Absolutely no reported issues.

Good.

While it's true they might be related i'm more inclined to treat these two (scrub and receive) as separate issues.

Also, until i reproduce this on my system, i'm kind of flying blind here and you have to be our eyes for the time being: every bit of information can be useful.

1. Does "Absolutely no reported issues" mean that the scrub did not report any error and took the usual amount of time? Which is ~1h?

2. Same behavior on ZoL 0.7.0-rc3: hung process with 100% utilization and no exit. Ok, does that also mean same stack?

**Comment by tannerdsilva** [comment]
> While it's true they might be related i'm more inclined to treat these two (scrub and receive) as separate issues.

With this in mind, I tried a **scrub** on ZoL 0.7.0-rc3. Thankfully, this issue has been resolved. I have working scrubs on my `tank` pool again. (cc: @mrenouf)

Given that scrubs are working as expected, I'm inclined to think `zfs send | zfs recv` will work as well (assuming they're both on 0.7.0, not the asymmetric 0.6.5.9 send | 0.7.0 recv like I tried yesterday. I think 0.6.5.9 was probably piping bad data).

Don't have the time to test send | recv on 0.7.0 at the moment, but I will open a new issue if that problem persists.

Thank you @loli10K for your helpful and prompt responses.

**Comment by loli10K** [comment]
@tannerdsilva i hope you don't take this the wrong way but i kind of have mixed feelings about this: first off i'm glad that things worked out for you using 0.7.0 (but keep in mind that upgrading the pool will probably prevent you from using it on FreeBSD).
On the other hand we still have an unresolved bug that is apparently affecting the stable branch and that we don't know (yet) how to reproduce. Upgrading to 0.7.0 may have fixed your pool, but we don't know "how" and "why". From my selfish point of view we still had a lot of work to do to extact meaningful information on how this issue behaves: 

1. compare "dbgmsg" data from scrubs on FreeBSD and ZoL 0.6.5/0.7.0
2. bring back the pool on ZoL 0.6.5 and see if it's still broken
3. dump the DTL entries via `zdb`
4. trace and compare the IO being done during a scrub on FreeBSD and ZoL 0.6.5/0.7.0

It would be nice if @mrenouf could post more info about his setup.

**Comment by tannerdsilva** [comment]
@loli10K Glad to help in any way I can, anything you need.

I mentioned this in tidbits here and there, but its worth repeating to make sure we're on the same page: there are two unique characteristics with this "problematic dataset(s)" I've identified that are not present with any other sets in my pool. If you are still trying to reproduce the issue on your end, take this into consideration.
1. The dataset (and children) is simply a backup of another pool on my system. The problematic dataset that now sits on `tank` has been built and **incrementally refreshed** using `send | recv` from the source pool.
2. This is the only dataset on my pool that contains clones as children.

I'll get back to you on the info you've requested.

**Comment by mrenouf** [comment]
I have the history file, I haven't had a chance to post it. Will do that tonight.
Unfortunately I couldn't keep the pool around in a questionable state, I've had to recreate it from backup on ZoL.

A brief summary of my experience:

- Created a 5 disk raidz1 pool on FreeNAS "Corral", (FreeBSD 11)
- Used for about a month, no issues
- Exported and imported on Ubuntu Server 16.04.2
- Hardware failure takes out one drive. A replacement is installed.
- On 'zpool replace' of the disk, the following resilver  ran for a very short time (~minutes), then the pool was claimed to be "healthy".
- Subsequent scrubs yielded (not suprisingly) tons of checksum errors, and the scrub would also abort not long into the process.
- I then booted back into FreeNAS, and performed a scub which ran to completion.
- Booting back into Ubuntu, a full backup was made using rsync onto external drives (this took ages but completed any issues)

**Comment by loli10K** [comment]
@mrenouf thank you for taking the time to share this information. This is on ZoL 0.6.5.X right?

**Comment by mrenouf** [comment]
This is Ubuntu 16.04.2, out of the box. 
http://packages.ubuntu.com/yakkety/zfsutils-linux

The package is version: 0.6.5.8-0ubuntu4

**Comment by mrenouf** [comment]
PS: I used rsync under Linux onto ext4 because I was worried about the zfs send/receive issues described earlier, and the risk of perpetuating some kind of incompatibility by backing up onto ZFS created by FreeBSD. Maybe this was not necessary, but I didn't want to risk it (it took about 4 days total to backup/restore).

**Comment by loli10K** [comment]
@mrenouf given the circumstances i would have probably done the same. Do you remember scrubbing the pool before replacing the disk, either manually or via some kind of job scheduler?

**Comment by tannerdsilva** [comment]
@loli10K Sorry for the delay, finals week at school has me abnormally occupied.

Taking a brief look at `dbgmesg` on 0.7.0....the big thing that stands out to me are the lines that look like
```
scanned dataset XXXX (DATASET_PATH) with min=# max=#########; pausing=0
```
Search with grep reveals that min is **always** 1 in 0.7.0, except for the at the origin dataset, where it is 0. Compare [this](https://gist.github.com/tannerdsilva/36a93dd6eed639c9bbdeb335b4934453#file-dbmesg0-7-0) to the `dbgmesg` I originally posted, notice that min is never 1.

[This](https://gist.github.com/tannerdsilva/631b0276589cf9ea31a2fbea235c5e39#file-zdb) is what zdb looks like (I upgraded tank while running 0.7.0, not comfortable letting 0.6.9.5 write to tank so I figured it was ok to pull the trigger).

> - bring back the pool on ZoL 0.6.5 and see if it's still broken
...
> - trace and compare the IO being done during a scrub on FreeBSD and ZoL 0.6.5/0.7.0

I'll get back to you on this, my box is headless and needs to be moved to a monitor so I can boot into the 0.6.5 environment I forked.

**Comment by mrenouf** [comment]
Sorry it took so long!
Here's my zpool history, covers from creation up until the events I listed earlier.

https://gist.github.com/mrenouf/268611c1377268648711f080a0e06b62

**Comment by tannerdsilva** [comment]
@loli10K looks like scrubs are also working on 0.6.5 now. Very strange.

**Comment by loli10K** [comment]
> looks like scrubs are also working on 0.6.5 now. Very strange.

@tannerdsilva this is, again, both good and bad: it seems we lost the ability to reproduce the issue but, on the other hand, we have another piece of the puzzle and also you may have discovered a possible workaround for anyone that doesn't want to rebuild their pool.

> Search with grep reveals that min is always 1 in 0.7.0, except for the origin dataset, where it is 0. Compare this to the dbgmesg I originally posted, notice that min is never 1.

I see several lines where "min=1" in 0.6.5 output, which are probably datasets without snapshots (or the oldest dataset snapshot). Unfortunately the output from 0.7.0 seems incomplete so we can't really compare the two.

Until we have a case we can safely investigate i'd like to keep this open, if you don't mind.

**Comment by tannerdsilva** [comment]
@loli10K I agree, the issue should stay open. I have completed scrubs, but I'm also looking at a 100% zed process (still on 0.6.5). Appears to be stuck. This process doesn't seem to be scrub related, as it is present even after a reboot (note the uptime).

```
top - 16:49:41 up 1 min,  1 user,  load average: 0.69, 0.21, 0.08
Tasks: 368 total,   2 running, 366 sleeping,   0 stopped,   0 zombie
%Cpu(s):  3.9 us,  5.6 sy,  0.0 ni, 89.0 id,  1.4 wa,  0.0 hi,  0.1 si,  0.0 st
KiB Mem:  16479676 total,   340028 used, 16139648 free,     1700 buffers
KiB Swap:   976892 total,        0 used,   976892 free.    76432 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND    
 4419 root      20   0   33556   1772   1544 R  97.9  0.0   0:48.12 zed        
    1 root      20   0   28840   5328   3272 S   0.0  0.0   0:01.35 systemd    
    2 root      20   0       0      0      0 S   0.0  0.0   0:00.02 kthreadd   
```

**Comment by tannerdsilva** [comment]
These problems also persist when the (supposed) problematic dataset I mentioned earlier is removed from `tank`. So it seems like we're back to square one.

**Comment by loli10K** [comment]
> but I'm also looking at a 100% zed process (still on 0.6.5)

This could be the same issue reported in https://github.com/zfsonlinux/zfs/issues/4720: @tannerdsilva can you run `zpool events` and see if it reports some kind of error, like "_internal error: Bad file descriptor_"? If it does, we have the opportunity to debug and fix that bug too.

**Comment by tannerdsilva** [comment]
@loli10K yep, same issue. Exactly as you described.

**Comment by tannerdsilva** [comment]
I think I'm having issues with zvols. I have a few zvols for my VM's on another pool, worked fine on 0.7.0, but I'm not even getting a zvol directory in /dev on 0.6.5. Also, I just realized the problematic dataset I previously mentioned uniquely contained volumes (it no longer exists on my system, I'd guess this is why the scrubs are working again).

**Comment by tannerdsilva** [comment]
I'd bet this issue can be reproduced if I make a volume, and then make clones (with diffs) of said volume. I'll give it a shot.

**Comment by loli10K** [comment]
@tannerdsilva nice, well spotted, i'm going to try that on my bsd11 test pool when i have a little time.

Regarding the zed issue, can you `strace -f -p $pidof_zed` and see if it's doing many syscall (and which ones)? Also try `gdb -q -batch -ex "attach $pidof_zed" -ex "thread apply all bt` and `lsof -p $pidof_zed`, and post a gistfile, thanks!

If we are in the same situation as described in https://github.com/zfsonlinux/zfs/issues/4720#issuecomment-245532566 (many ZFS_IOC_EVENTS_NEXT ioctl returning EBADF) and you feel comfortable running SystemTap on your box, you could try running the following stap script which will (hopefully) shed some light on why the kernel is returning EBADF:

contents of "whyebadf.stp":
```c
#!/usr/bin/env stap

probe
module("zfs").function("zfs_zevent_wait").return,
module("zfs").function("zfs_zevent_fd_hold").return,
module("zfs").function("zfsdev_get_state").return,
module("zfs").function("zfs_zevent_next").return
{
   printf("%lu %s <- %s %s\n", gettimeofday_s(), symname(caller_addr()), ppfunc(), $$return$);
}
```

Run `stap -v ./whyebadf.stp` until you get enough information, probably just a couple of seconds if zed is doing a lot of ioctls.

**Comment by tannerdsilva** [comment]
@loli10K Running `strace` on the `zed` process returns hundreds of identical lines:
```
ioctl(5, 0x5a81, 0x7ffecee5ee40)        = -1 EBADF (Bad file descriptor)
```
[Output of the gdb and lsof commands](https://gist.github.com/tannerdsilva/75c9fa204a12a86eae34d991cb94655d#file-command-output)

[stap output](https://gist.github.com/tannerdsilva/3b379fc93be8ae7a13ff88d8d92d02e5#file-stap_output)

**Comment by tannerdsilva** [comment]
The volumes I create in 0.7.0 are not showing up when I do an import on 0.6.5. So if the volumes (as I suspect) are the root cause of the scrub issues, it is not specific to BSD. 

Make a pool in 0.7.0, make a volume and occupy with data. Move back to 0.6.5 stable and /dev/zvol should not be present after the pool is imported.

**Comment by loli10K** [comment]
Good news, i've reproduced the original issue (interrupted scrub) on a small raidz pool using FreeNAS-9.10.2-U4: now the fun part is actually understanding what is causing it and finally come up with a fix.

**Comment by loli10K** [comment]
Bad news, this bug is also reproducible on 0.7.0-rc4: pools created with newer software versions (and the proper features flags disabled) cannot be scrubbed successfully on 0.6.5.X.

**Comment by tannerdsilva** [comment]
When I take my rc-3 created pools and import them to 0.6.5.9, I see other issues as well. Ex: after a seemingly successful import, one of my pools wont show up in `zpool status`. And as stated earlier, zvols are causing headaches.

**Comment by loli10K** [comment]
@tannerdsilva given the nature of the original issue (interrupted scrub) and the fact we can reproduce it without involving ZVOLs (or RAIDZ, as a matter of facts) i'd say those other issues have nothing to do with the first one.

**Comment by loli10K** [comment]
This should be fixed in https://github.com/zfsonlinux/zfs/commit/94d353a0bf5ee34adb69fe018528b25577457e7a.

**Comment by eblau** [comment]
I'm having what appears to be the same zfs receive issue reported in this issue. I'm trying to receive a stream created on FreeNAS 9.10.2. When I run the zfs receive on the zfsonlinux side running 0.6.5.9, the zfs receive seems to complete (I see the dataset show up in zfs list), but the zfs receive command hangs "forever" taking 100% CPU and the dataset is not actually mounted successfully.

This issue is completely reproducible for me with this one dataset coming from FreeNAS. Other datasets can be received with no issues. I've checked the zpool feature flags and everything agrees between the two sides so I'm stumped.

Was the zfs receive issue reported here ever resolved? I see that the scrub issue appears to be resolved but what about the other issue?

**Comment by gmelikov** [comment]
@eblai, as you can see in last comment here, it's already fixed in 0.6.5.11.

**Comment by eblau** [comment]
@gmelikov I read all the comments in this issue. loli10K says:

> While it's true they might be related i'm more inclined to treat these two (scrub and receive) as separate issues.

and I don't see any confirmation that the zfs receive issue was fixed. The fix in 94d353a also refers to scrub and not to the receive issue which is why I asked for confirmation.

**Comment by behlendorf** [comment]
@eblau it's possible your hitting issue #6616.  There was an upstream OpenZFS change which subtly broke the stream compatibility on the send side.  This was caught after the change was pulled in for Linux and it's likely FreeNAS 9.10.2 has the same issue.  It will only cause issues when sending to older versions of the software, which is how it slipped through.

**Comment by eblau** [comment]
@behlendorf Thanks for the pointer. I don't have control over the FreeNAS side, but I can upgrade the Linux side to 0.7.2 and see if it solves the issue. Thanks again!
